### PR TITLE
Non-"CRUD-er" DSRs

### LINKED ISSUE #5012: Internal Server Error when missing key in POST to /deliveryservice_requests
## I'm submitting a ...
-  bug report

## Traffic Control components affected ...
-  Traffic Ops

## Current behavior:
When passing a request that's missing the `deliveryservice` property to `/deliveryservice_request`'s POST handler, the request's goroutine will segfault and panic.

## Expected / new behavior:
Invalid requests should not be capable of causing a server-side segfault.

## Minimal reproduction of the problem with instructions:
POST to `/deliveryservice_request` with a request body not containing a `deliveryservice` property (probably `"deliveryservice": null` works too, but I haven't checked). Simplest body to reproduce the error is just `{}`.

## Anything else:
<details>
<summary>Full HTTP stack</summary>

```http
POST /api/2.0/deliveryservice_requests HTTP/1.1
User-Agent: python-requests/2.22.0
Accept-Encoding: gzip, deflate
Accept: */*
Connection: keep-alive
Cookie: mojolicious=eyJhdXRoX2RhdGEiOiJhZG1pbiIsImV4cGlyZXMiOjE1OTkxODIwNjQsImJ5IjoidHJhZmZpY2NvbnRyb2wtZ28tdG9jb29raWUifQ--fd12037b48601a4f29a16fd06a9881c19be9456b
Content-Length: 2

{}
```

```http
HTTP/1.1 500 Internal Server Error
Access-Control-Allow-Credentials: true
Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Set-Cookie, Cookie
Access-Control-Allow-Methods: POST,GET,OPTIONS,PUT,DELETE
Access-Control-Allow-Origin: *
Content-Encoding: gzip
Content-Type: application/json
Set-Cookie: mojolicious=eyJhdXRoX2RhdGEiOiJhZG1pbiIsImV4cGlyZXMiOjE1OTkxNjQwNjQsImJ5IjoidHJhZmZpY2NvbnRyb2wtZ28tdG9jb29raWUifQ--de133c58fd1afe4b253fe6d920be7002dfdb4513; Path=/; Expires=Thu, 03 Sep 2020 20:14:24 GMT; Max-Age=3600; HttpOnly
Vary: Accept-Encoding
Whole-Content-Sha512: /HT8WHLIsAIPgTPjZAa4GmpUwYtwph3fcU1//Cr7W4rkL1lKRW3kUEDTIz1DoA3/zUm3oa0elunc5VHcism6Qg==
X-Server-Name: traffic_ops_golang/
Date: Thu, 03 Sep 2020 19:14:24 GMT
Content-Length: 83

{
	"alerts": [
		{
			"text": "Internal Server Error",
			"level": "error"
		}
	]
}
```

</details>

<details>
<summary>Traffic Ops Logs</summary>

```
INFO: routing.go:249: 2020-09-03T19:05:31.140864324Z: POST /api/2.0/deliveryservice_requests? handling (reqid 2750)
DEBUG: plugin.go:190: 2020-09-03T19:05:31.140914016Z: DEBUG plugins.OnRequest calling 0 plugins
ERROR: api.go:222: 2020-09-03T19:05:31.151730185Z: 192.168.176.1:41492 panic: (err: runtime error: invalid memory address or nil pointer dereference) stacktrace:
goroutine 21149 [running]:
github.com/apache/trafficcontrol/lib/go-util.Stacktrace(0xc000fed398, 0xce4900, 0x15d8b00)
	/tmp/go/src/github.com/apache/trafficcontrol/lib/go-util/util.go:32 +0x9d
github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/routing/middleware.WrapPanicRecover.func1.1(0xf74760, 0xc0016a17d0, 0xc001368400)
	/tmp/go/src/github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/routing/middleware/wrappers.go:152 +0x54
panic(0xce4900, 0x15d8b00)
	/usr/local/go/src/runtime/panic.go:969 +0x166
github.com/apache/trafficcontrol/lib/go-tc.(*DeliveryServiceNullable).Validate(0x0, 0xc000f98100, 0x3, 0x4)
	/tmp/go/src/github.com/apache/trafficcontrol/lib/go-tc/deliveryservices.go:543 +0x263
github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/deliveryservice/request.(*TODeliveryServiceRequest).Validate(0xc000eda000, 0xda2260, 0xc000eda000)
	/tmp/go/src/github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/deliveryservice/request/validate.go:63 +0x43e
github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api.decodeAndValidateRequestBody(0xc001368400, 0x7f9b62f5e2b8, 0xc000eda000, 0x0, 0x0)
	/tmp/go/src/github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api/shared_handlers.go:115 +0x13d
github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api.CreateHandler.func1(0xf74760, 0xc0016a17d0, 0xc001368400)
	/tmp/go/src/github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api/shared_handlers.go:615 +0x10da
github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/routing/middleware.AuthBase.GetWrapper.func1.1(0xf74760, 0xc0016a17d0, 0xc001368400)
	/tmp/go/src/github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/routing/middleware/wrappers.go:107 +0x276
github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/routing/middleware.WrapPanicRecover.func1(0xf74760, 0xc0016a17d0, 0xc001368400)
	/tmp/go/src/github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/routing/middleware/wrappers.go:156 +0x79
github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/routing/middleware.WrapHeaders.func1(0xf775e0, 0xc00333c420, 0xc001368400)
	/tmp/go/src/github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/routing/middleware/wrappers.go:136 +0x643
net/http.HandlerFunc.ServeHTTP(0xc00047e9b0, 0xf775e0, 0xc00333c420, 0xc001368400)
	/usr/local/go/src/net/http/server.go:2012 +0x44
net/http.(*timeoutHandler).ServeHTTP.func1(0xc00333c540, 0xc000191c40, 0xc00333c420, 0xc001368400, 0xc000114360)
	/usr/local/go/src/net/http/server.go:3238 +0x7f
created by net/http.(*timeoutHandler).ServeHTTP
	/usr/local/go/src/net/http/server.go:3232 +0x226
```

</details>

### LINKED ISSUE #5014: Creating a Delivery Service Request with an ID fails
## I'm submitting a ...
-  bug report

## Traffic Control components affected ...
-  Traffic Ops

## Current behavior:
Because of the way workflow transitions are verified, creating a DSR with any non-null value in the `id` property will cause the request to fail. If `id` was an integer, though, it will give back the error-level alert: `sql: no rows in result set`, which is both cryptic and - looking at the code - unnecessary.

## Expected / new behavior:
If IDs have special meaning to requests, they should be documented - but it's more likely that the intended behaviour is just to ignore the ID, since it can't reasonably be known before the object is created.

## Minimal reproduction of the problem with instructions:
Submit a POST request to `/deliveryservice_requests` with an integral ID value. The simplest request body to reproduce the error is just `{"id": -1}` (which should work regardless of extant data).
## Anything else:
<details>
<summary>Full HTTP Stack</summary>

```http
POST /api/2.0/deliveryservice_requests HTTP/1.1
User-Agent: python-requests/2.22.0
Accept-Encoding: gzip, deflate
Accept: */*
Connection: keep-alive
Cookie: mojolicious=eyJhdXRoX2RhdGEiOiJhZG1pbiIsImV4cGlyZXMiOjE1OTkxOTI1ODIsImJ5IjoidHJhZmZpY2NvbnRyb2wtZ28tdG9jb29raWUifQ--2d0e7a169ef17f538d58830ee8451e2113562119
Content-Length: 9

{"id": 2}
```
```http
HTTP/1.1 400 Bad Request
Access-Control-Allow-Credentials: true
Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Set-Cookie, Cookie
Access-Control-Allow-Methods: POST,GET,OPTIONS,PUT,DELETE
Access-Control-Allow-Origin: *
Content-Encoding: gzip
Content-Type: application/json
Set-Cookie: mojolicious=eyJhdXRoX2RhdGEiOiJhZG1pbiIsImV4cGlyZXMiOjE1OTkxNzQ1ODIsImJ5IjoidHJhZmZpY2NvbnRyb2wtZ28tdG9jb29raWUifQ--88ac7381f3f3d92bb3fd8dc8975be3896a2acc3d; Path=/; Expires=Thu, 03 Sep 2020 23:09:42 GMT; Max-Age=3600; HttpOnly
Vary: Accept-Encoding
Whole-Content-Sha512: kLdrEUgBgalHQ/z01sSUnd5k/TCwjNOnjY6tZWxY88JIhEsiOPoACLxUIGPad+ErIRJ+l2mpC5g8/89l34o0jA==
X-Server-Name: traffic_ops_golang/
Date: Thu, 03 Sep 2020 22:09:42 GMT
Content-Length: 88

{
	"alerts": [
		{
			"text": "sql: no rows in result set",
			"level": "error"
		}
	]
}
```

</details>

### LINKED ISSUE #5015: Creating a Delivery Service Request has issues verifying tenant access
## I'm submitting a ...
-  bug report

## Traffic Control components affected ...
-  Traffic Ops

## Current behavior:
When submitting a POST request to `/deliveryservice_requests`, if the request body's `deliveryservice` object has no `tenantId` property (or probably also `null` but I haven't checked) the handler will return a 500 Internal Server Error to the client.

Also, the Tenant ID checked for user access is whatever the user submitted, - **not** the actual Tenant of the Delivery Service. In this way, a `change` or `delete` request can be submitted by a user who does not have permission to modify that Delivery Service. So a user with a Tenant ID of 3, which is a child of Tenant 1 can legally submit a request for any Delivery Service accessible to 1 by simply replacing the `tentantId` property of the `deliveryservice` object with their own Tenant's ID.

## Expected / new behavior:
If a property is required, then omitting it should cause a 400 Bad Request response with an appropriate `alerts` entry describing the problem.

A resource's actual Tenancy should be checked - returning a 403 Forbidden response in the event that access is not allowed - instead of whatever the client asks to be checked.

## Minimal reproduction of the problem with instructions:
Make a POST request to `/deliveryservice_requests` containing a syntactically valid DSR, but one wherein the `tenantId` of the `deliveryservice` object is undefined.

Make a POST request to `/deliveryservice_requests` containing a `deliveryservice` inaccessible to the requesting Tenant, but with the `deliveryservice`'s `tentantId` modified to an accessible value.

### LINKED ISSUE #5033: Internal Server Error when assigning a non-existent DSR
## I'm submitting a ...
-  bug report

## Traffic Control components affected ...
-  Traffic Ops

## Current behavior:
Trying to assign a Delivery Service Request which does not exist to any user will result in an Internal Server error.

## Expected / new behavior:
If a resource is not found, the API should communicate that with the `404 Not Found` HTTP status code.

## Minimal reproduction of the problem with instructions:
Given that no Delivery Service Request exists with the ID 1, attempt to assign Delivery Service Request 1 to any (existing) user. I used my own user (who had the ID 9 in the following example) and it looks like this:

### Request
```http
PUT /api/2.0/deliveryservice_requests/1/assign HTTP/1.1
User-Agent: python-requests/2.22.0
Accept-Encoding: gzip, deflate
Accept: */*
Connection: keep-alive
Cookie: mojolicious=...
Content-Length: 15

{"assigneeId": 9}
```

(Note also that the documentation specifies a required "id" property of the request body - that is not correct. The DSR to be assigned is fully identified by the request URI.)

### Response
```http
HTTP/1.1 500 Internal Server Error
Content-Encoding: gzip
Content-Type: application/json
Set-Cookie: mojolicious=...
Vary: Accept-Encoding
X-Server-Name: traffic_ops_golang/
Date: Tue, 15 Sep 2020 23:34:43 GMT
Content-Length: 83

{
	"alerts": [
		{
			"text": "Internal Server Error",
			"level": "error"
		}
	]
}
```
## Anything else:
From the stack trace I found:
```
INFO: routing.go:249: 2020-09-15T23:26:42.644209507Z: PUT /api/2.0/deliveryservice_requests/1/assign? handling (reqid 33624)
DEBUG: plugin.go:190: 2020-09-15T23:26:42.644259514Z: DEBUG plugins.OnRequest calling 0 plugins
ERROR: api.go:209: 2020-09-15T23:26:42.651463829Z: 172.20.0.1:47534 dsr assignment querying existing: sql: no rows in result set
DEBUG: api.go:214: 2020-09-15T23:26:42.651514639Z: Internal Server Error
172.20.0.1:47534 - admin [15/Sep/2020:23:26:42 +0000] "PUT /api/2.0/deliveryservice_requests/1/assign? HTTP/1.1" 500 83 7 "python-requests/2.22.0" 2703160290 NON_IMS
INFO: routing.go:252: 2020-09-15T23:26:42.652031369Z: PUT /api/2.0/deliveryservice_requests/1/assign? handled (reqid 33624) in 7.774269ms
```
which suggests that a `database/sql.ErrNoRows` isn't being handled somewhere.

### PR DESCRIPTION
**Pull request by ocket8888**
## What does this PR (Pull Request) do?
- [x] This PR fixes #5033 and fixes #5015 and fixes #5014 and fixes #5012

This also, importantly, rewrites DSRs to not use the "CRUD-er", which will allow their structure to vary between API versions. This is necessary for upcoming changes to the endpoint which were previously included in the now-obsolete #5071 .

## Which Traffic Control components are affected by this PR?
- Documentation
- Traffic Ops Go Client
- Traffic Ops

## What is the best way to verify this PR?
Run all associated tests, verify that the reproduction methods for each fixed bug no longer reproduce the respective bug.

## If this is a bug fix, what versions of Traffic Control are affected?
- master
- 5.0.0
- 4.1.1

## The following criteria are ALL met by this PR
- [x] This PR includes tests
- [x] This PR includes documentation
- [x] This PR includes an update to CHANGELOG.md
- [x] This PR includes any and all required license headers
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY**

### CODE REVIEW

**Review by zrhoffman — COMMENTED** [review]
API tests fail. Most of them look like they are related to alert text not matching exactly, but this one in `UpdateTestDeliveryServiceRequestsWithHeaders()` looks different:

```go
2021-02-02T23:44:57.0834003Z     deliveryservice_requests_test.go:101: CreateTestDeliveryServiceRequests
2021-02-02T23:44:57.0934609Z     deliveryservice_requests_test.go:105: Response:  {[{Delivery Service request created success}]}
2021-02-02T23:44:57.1057269Z     deliveryservice_requests_test.go:310: Response:  {[{Delivery Service Request #10 updated success}]}
2021-02-02T23:44:57.1165189Z     deliveryservice_requests_test.go:71: Expected error about precondition failed, but got none
2021-02-02T23:44:57.1166109Z     deliveryservice_requests_test.go:74: Expected status code 412, got 200
2021-02-02T23:44:57.1291805Z     deliveryservice_requests_test.go:71: Expected error about precondition failed, but got none
2021-02-02T23:44:57.1292734Z     deliveryservice_requests_test.go:74: Expected status code 412, got 200
2021-02-02T23:44:57.1428547Z     deliveryservice_requests_test.go:339: Response:  {[{Delivery Service Request #10 deleted success}]}
2021-02-02T23:44:57.3128663Z     types_test.go:208: ---- DeleteTestTypes ----
2021-02-02T23:44:57.4984572Z     --- FAIL: TestDeliveryServiceRequests (0.60s)
```

**Comment by ocket8888** [comment]
Yeah, I kinda expected that. Full disclosure: I didn't run the tests. Wanted the CI actions to do it for me.

I'm on it.

File: docs/source/glossary.rst [hunk 1]
```diff
@@ -115,7 +115,9 @@ Glossary
 		.. seealso:: See :ref:`delivery-services` for a more in-depth explanation of :dfn:`Delivery Services`.
 
 	Delivery Service Request
+	Delivery Service REquests
```
Comment (mitchell852) [inline docs/source/glossary.rst:118]: typo

**Comment by mitchell852** [comment]
For some reason I am unable to change the status. Example:

```
Request URL: https://localhost:8442/api/4.0/deliveryservice_requests/4641/status
Request Method: PUT
Request Payload: {status: "submitted"}
Status Code: 200 OK
```

```
Request URL: https://localhost:8442/api/4.0/deliveryservice_requests/4641/status
Request Method: GET
Status Code: 200 OK

{"response":"draft"}
```

**Comment by ocket8888** [comment]
> For some reason I am unable to change the status. Example:

Looks like a bad rebase from the old PR. Should be working now.

**Comment by mitchell852** [comment]
@ocket8888 - looks like these query params are no longer respected:

```
GET api/4.0/deliveryservice_requests?orderby=createdAt&sortOrder=desc
```

**Comment by mitchell852** [comment]
This doesnt seem right. here's the change log message for updating a DSR:

```
Updated Delivery Service Request of type update for Delivery Service ''
```

**Comment by ocket8888** [comment]
I was able to get `orderby` and `sortOrder` to work. I even tried specifically ordering by `createdAt`. Worked fine.

**Comment by mitchell852** [comment]
> I was able to get `orderby` and `sortOrder` to work. I even tried specifically ordering by `createdAt`. Worked fine.

yeah i can't get the `sortOrder` to stick when i do GET /api/4.0/deliveryservice_requests?orderby=createdAt&sortOrder=desc&limit=10 in your branch, i get this query:

```
SELECT
	a.username AS author,
	e.username AS lastEditedBy,
	s.username AS assignee,
	r.assignee_id,
	r.author_id,
	r.change_type,
	r.created_at,
	r.id,
	r.last_edited_by_id,
	r.last_updated,
	r.deliveryservice,
	r.status
FROM deliveryservice_request r
JOIN tm_user a ON r.author_id = a.id
LEFT OUTER JOIN tm_user s ON r.assignee_id = s.id
LEFT OUTER JOIN tm_user e ON r.last_edited_by_id = e.id
WHERE (
	CAST(r.deliveryservice->>'tenantId' AS BIGINT) = ANY(CAST(:accessibleTenants AS BIGINT[]))
)
ORDER BY r.created_at
LIMIT 10
```

^^ see? no DESC attached to order by in the query but in master i get it.

**Comment by ocket8888** [comment]
Actually it looks like the ordering was only working for fields other than `createdAt`, and when I tested that one it was only ordered by pure coincidence. Should be fixed now. Also added documentation for the sorting and pagination query parameters.

**Review by mitchell852 — APPROVED** [review]
looks good. everything works as it did before as far as i can tell.

**Pull request merged**